### ISSUE TITLE
Extend native scorer beyond MSE to support all NEAT-AI cost functions

### ISSUE THREAD

**Issue by nleck** [stSoftwareAU/NEAT-AI-scorer#119]
We've added new Costs https://github.com/stSoftwareAU/NEAT-AI/blob/Develop/src/Costs.ts 

Please make sure the native scorer handles the costs that it is possible and raise the supporting GH issues in the correct repositories to support passing the costs ( which can be supported by rust). 

Costs that can be supported by native Rust will be faster. Where we can support the cost function in Rust make sure the type script off loads to the native rust application for faster performnce. 

This will mean changes in NEAT-AI-core and NEAT-AI please raise GH issues as needed.

## Current Understanding

The native Rust scorer currently fuses MSE only. The TypeScript `Costs.ts` exposes seven built-in cost functions (CrossEntropy, MSE, MAE, MAPE, MSLE, HINGE, CategoricalError) plus a user-registered custom-cost extension point. The work is to extend `rust_scorer` (and `neat-core`) so the TS `NEAT-AI` layer can off-load any of the seven built-ins to Rust, define the IPC contract for cost selection, and raise the matching follow-up issues in `NEAT-AI-core` and `NEAT-AI`.

### Accepted scope so far
- Rust supports **all seven** built-in costs in this round: CrossEntropy, MSE, MAE, MAPE, MSLE, HINGE, CategoricalError.
- IPC: TS selects the cost via a new `--cost <name>` CLI flag on the Rust binary. **No environment-variable override** — keeps the surface small and avoids extra permission scope.
- Unsupported-cost behaviour: **hard error from Rust**. TS must not request a cost name the binary does not implement.
- GPU policy (KISS): extend the GPU kernel to a new cost **only if** a quick bench shows it is significantly faster than the CPU+PGO path; otherwise the cost stays CPU-only in Rust. No blanket GPU rewrite.
- Acceptance bar: **numerical parity tests vs the TS reference** for every supported cost, within a small fixed tolerance.
- Custom (user-registered) TS costs remain TS-only — out of scope for the Rust off-load path.
- Raise follow-up issues in `NEAT-AI-core` (cost name plumbing into the scorer call) and `NEAT-AI` (selecting the cost name when invoking the native scorer).

### Open questions
- None.

### Assumptions
- "Significantly faster" for the GPU gating decision is left to engineering judgement at implementation time (rule of thumb: clear, repeatable win on the standard bench — not a marginal speedup); if a sharper threshold is needed later it can be set in the GPU follow-up issue.
- Parity tolerance is a small fixed epsilon chosen per cost (e.g. 1e-9 for smooth costs, larger for log-based costs where TS/Rust may differ in last-bit rounding); finalised in code review.
- GPU kernel extensions for any cost that does not clearly win the bench will be tracked as separate follow-up issues, not blockers for this issue.

### Related open issues
- #77 — "Right sizing creature scoring": adjacent performance concern, no direct overlap with cost selection.
- #87 — Linux + NVIDIA GPU benchmark follow-up: independent; any new GPU cost kernel from this issue would eventually be benchmarked there but does not block this work.

**Comment by stservice** [comment]
## Grill-Me Round 1

**TL;DR** — The Rust scorer only does MSE today. The TS `Costs.ts` has 7 built-ins (CrossEntropy, MSE, MAE, MAPE, MSLE, HINGE, CategoricalError). Need your call on scope, IPC, and fallback before I converge.

### Understanding
Goal: extend `rust_scorer` (and `neat-core`) so the TS `NEAT-AI` layer can off-load any supported cost to Rust for speed, and raise the matching follow-up issues in `NEAT-AI-core` / `NEAT-AI`. Custom (user-registered) TS costs are assumed out of scope for the Rust path.

### Questions

1. Which built-in costs should Rust support in this first round? (tick all that apply)
   - [x] All seven (CrossEntropy, MSE, MAE, MAPE, MSLE, HINGE, CategoricalError)
   - [ ] Regression set only (MSE, MAE, MAPE, MSLE)
   - [ ] Classification set only (CrossEntropy, HINGE, CategoricalError)
   - [ ] MSE + CrossEntropy first, rest in follow-ups
   - [ ] other — please describe in a reply

2. How should TS tell the Rust binary which cost to use?
   - [x] New `--cost <name>` CLI flag (with env override `NEAT_SCORER_COST`)
   - [ ] Field on the creature JSON (e.g. `costName`)
   - [ ] Field on the training data / config JSON
   - [ ] other — please describe in a reply

3. What should happen when TS asks for a cost Rust does not implement?
   - [x] Hard error from Rust — TS must not request unsupported costs
   - [ ] Rust returns "unsupported" sentinel; TS falls back to its own implementation
   - [ ] TS gates off-load by name and never asks Rust for unsupported costs
   - [ ] other — please describe in a reply

4. Is GPU support for the new costs in scope for this issue?
   - [ ] No — CPU off-load only; GPU kernels stay MSE-only for now (raise separate follow-up)
   - [ ] Yes — extend the batched GPU kernel to every cost we add
   - [x] Only for the costs that already beat CPU+PGO in a quick bench
   - [ ] other — please describe in a reply

5. Acceptance bar for "Rust matches TS"?
   - [x] Numerical parity tests vs TS reference for every supported cost (small fixed tolerance)
   - [ ] Spot-check tests on a handful of fixtures, no per-element parity
   - [ ] Bench-only — must be at least as fast as TS, correctness via existing scorer tests
   - [ ] other — please describe in a reply

Tick the boxes that apply (the GitHub mobile app lets you tap each one), then add any free-form notes in a reply.

**⏳ Awaiting your reply.** The Vibe Coder is waiting for your response on this issue to continue grilling.

**Comment by nleck** [comment]
Q2) I don't think we need the environment variable, then we need additional permissions etc

**Comment by nleck** [comment]
q4) if GPU helps then use it ( significantly) then use the gpu else KISS

**Comment by VibeCoderST** [comment]
## Grill-Me — Ready for Next Phase

**TL;DR** — Rust scorer extends to all 7 built-in costs via a `--cost <name>` CLI flag, hard-errors on unsupported costs, uses GPU only where it clearly wins the bench, and is gated by numerical-parity tests vs the TS reference.

### Agreed scope
- **Costs**: CrossEntropy, MSE, MAE, MAPE, MSLE, HINGE, CategoricalError implemented in Rust (`rust_scorer` / `neat-core`).
- **IPC**: TS selects the cost via `--cost <name>` CLI flag. No env var (keeps permissions narrow).
- **Unsupported costs**: hard error from Rust.
- **GPU**: KISS — extend the GPU kernel to a cost only if it is significantly faster than CPU+PGO; otherwise CPU-only in Rust.
- **Acceptance**: numerical-parity tests vs the TS reference, small fixed tolerance per cost.
- **Out of scope**: user-registered (custom) TS costs stay TS-only.
- **Follow-ups**: raise issues in `NEAT-AI-core` (cost-name plumbing) and `NEAT-AI` (selecting cost name when invoking native scorer).

### Assumptions
- "Significantly faster" left to engineering judgement at bench time.
- Parity epsilon chosen per cost in code review.
- GPU kernels for costs that don't clearly win the bench are separate follow-ups, not blockers.

### Related open issues
- #77 — "Right sizing creature scoring": adjacent perf concern, no direct overlap.
- #87 — Linux + NVIDIA GPU benchmark follow-up: independent; new GPU cost kernels would benchmark there.

### How would you like to proceed?
- [ ] Apply the `planning` label
- [ ] Apply the `work-on` label

Tick the box for your choice, then open the issue on GitHub mobile, tap the labels icon, and toggle either `planning` or `work-on` to actually apply it.

**Comment by VibeCoderST** [comment]
Claimed by `VibeCoderST-1779545316804` on host `GRQ-23`

**Comment by VibeCoderST** [comment]
## Planning Complete

Broken down into 7 sub-issues across three repositories. The Rust side already has 6/7 cost helpers in `neat-core/src/loss.rs` (only `CategoricalError` is missing); the work is mostly CLI plumbing, dispatch wiring, parity tests, and the TS-side off-load contract.

### Sub-issues created

**NEAT-AI-core**
- stSoftwareAU/NEAT-AI-core#88 — Add `categorical_error_sum_batch_packed` (argmax misclassification rate).

**NEAT-AI-scorer (this repo)**
- #120 — Add `--cost <NAME>` CLI flag with hard-error on unknown cost.
- #121 — Wire `CostKind` dispatch through `stream_score` and `multi_score`; add `costName` to `ScoreResult`; gate GPU kernel to MSE only.
- #122 — Numerical parity tests vs TS reference for all seven costs.
- #123 — Documentation update (README, AGENTS.md, CLI help).
- #124 — Follow-up: bench non-MSE GPU kernels; raise per-cost issues for winners (or close as negative result).

**NEAT-AI**
- stSoftwareAU/NEAT-AI#2745 — Pass configured `costName` to `rust_scorer` via `--cost` from `RustScorerBridge`.

### Suggested implementation order

```mermaid
flowchart LR
    A[NEAT-AI-core#88 CategoricalError] --> C[#121 Dispatch wiring]
    B[#120 --cost CLI flag] --> C
    C --> D[#122 Parity tests]
    C --> E[#123 Docs]
    C --> F[#124 GPU bench follow-up]
    D --> G[NEAT-AI#2745 TS bridge]
```

### Dependency relationships

- `NEAT-AI-core#88` (CategoricalError in core) → `#121` (dispatch needs the function to exist).
- `#120` (CLI flag) → `#121` (dispatch needs the flag).
- `#121` (dispatch) → `#122`, `#123`, `#124` (parity tests, docs, GPU bench all need the real dispatch in place).
- `#121` + `#122` → `NEAT-AI#2745` (TS bridge passes `--cost`; should only land after correctness is proven).

Items with no edge between them (`NEAT-AI-core#88` vs `#120`; `#122` vs `#123` vs `#124`) can proceed in parallel.

### Assumptions and decisions

- **Cost name strings** follow the TS `BUILT_IN_COST_NAMES` exactly: `MSE`, `MAE`, `MAPE`, `MSLE`, `HINGE`, `CROSS_ENTROPY`, `CATEGORICAL_ERROR`. No aliases.
- **Default cost** remains `MSE` so the CLI is backward-compatible — omitting `--cost` matches today's behaviour.
- **No env-var override** for cost selection (issue brief is explicit; KISS).
- **GPU policy**: kernel stays MSE-only for now. `--gpu auto --cost <non-MSE>` falls back to CPU silently; `--gpu on --cost <non-MSE>` hard-errors. New kernels are gated behind the bench in #124.
- **Custom (user-registered) TS costs** stay TS-only. `NEAT-AI#2745` will bypass the Rust path entirely when a custom cost is configured.
- **Parity epsilons**: `1e-9` for smooth costs (MSE/MAE/MAPE/HINGE), `1e-6` for log-based (MSLE/CROSS_ENTROPY), exact equality for `CATEGORICAL_ERROR`. Finalised in #122 code review.

### Notes

This planning issue will be closed automatically. **Reopen it to request changes to the plan** before any of the sub-issues are picked up.

**Comment by VibeCoderST** [comment]
## Planning Complete

Planning complete. **5 sub-issue(s)** created:

- https://github.com/stSoftwareAU/NEAT-AI-scorer/issues/124
- https://github.com/stSoftwareAU/NEAT-AI-scorer/issues/123
- https://github.com/stSoftwareAU/NEAT-AI-scorer/issues/122
- https://github.com/stSoftwareAU/NEAT-AI-scorer/issues/121
- https://github.com/stSoftwareAU/NEAT-AI-scorer/issues/120

---
🤖 Processed by: VibeCoderST

**Comment by VibeCoderST** [comment]
Planning complete — sub-issues created.
